packages/jaeger-ui/src/model/ddg/Graph.tsx

```diff
@@ -121,33 +123,59 @@ export default class Graph {
       .join('____');
   };
 
-  public getVisible = (visEncoding?: string): { edges: TEdge[]; vertices: TDdgVertex[] } => {
-    const edges: Set<TEdge> = new Set();
-    const vertices: Set<TDdgVertex> = new Set();
-    const pathElems =
-      visEncoding == null
-        ? ([] as PathElem[]).concat(
-            this.distanceToPathElems.get(-2) || [],
-            this.distanceToPathElems.get(-1) || [],
-            this.distanceToPathElems.get(0) || [],
-            this.distanceToPathElems.get(1) || [],
-            this.distanceToPathElems.get(2) || []
-          )
-        : decode(visEncoding)
-            .map(visIdx => this.visIdxToPathElem[visIdx])
-            .filter(Boolean);
-
-    pathElems.forEach(pathElem => {
-      const edge = this.pathElemToEdge.get(pathElem);
-      if (edge) edges.add(edge);
-      const vertex = this.pathElemToVertex.get(pathElem);
-      if (vertex) vertices.add(vertex);
-      else throw new Error(`PathElem wasn't present in initial model: ${pathElem}`);
-    });
+  public getVisible: (visEncoding?: string) => { edges: TEdge[]; vertices: TDdgVertex[] } = memoize(10)(
+    (visEncoding?: string): { edges: TEdge[]; vertices: TDdgVertex[] } => {
+      const edges: Set<TEdge> = new Set();
+      const vertices: Set<TDdgVertex> = new Set();
+      const pathElems =
+        visEncoding == null
+          ? ([] as PathElem[]).concat(
+              this.distanceToPathElems.get(-2) || [],
+              this.distanceToPathElems.get(-1) || [],
+              this.distanceToPathElems.get(0) || [],
+              this.distanceToPathElems.get(1) || [],
+              this.distanceToPathElems.get(2) || []
+            )
+          : decode(visEncoding)
+              .map(visIdx => this.visIdxToPathElem[visIdx])
+              .filter(Boolean);
+
+      pathElems.forEach(pathElem => {
+        const edge = this.pathElemToEdge.get(pathElem);
+        if (edge) edges.add(edge);
+        const vertex = this.pathElemToVertex.get(pathElem);
+        if (vertex) vertices.add(vertex);
+        else throw new Error(`PathElem wasn't present in initial model: ${pathElem}`);
+      });
+
+      return {
+        edges: Array.from(edges),
+        vertices: Array.from(vertices),
+      };
+    }
+  );
+
+  public getVisibleUiFindMatches: (uiFind?: string, visEncoding?: string) => Set<TDdgVertex> = memoize(10)(
+    (uiFind?: string, visEncoding?: string): Set<TDdgVertex> => {
+      const vertexSet: Set<TDdgVertex> = new Set();
+      if (!uiFind) return vertexSet;
+
+      const uiFindArr = uiFind
+        .toLowerCase()
+        .split(' ')
+        .filter(Boolean);
+      const { vertices } = this.getVisible(visEncoding);
+      vertices.forEach(vertex => {
+        const { service, operation } = vertex;
+        const possibleMatch = `${service} ${operation}`.toLowerCase();
+        if (uiFindArr.some(str => possibleMatch.includes(str))) {
+          vertexSet.add(vertex);
+        }
+      });
```

Comment:
This sort of approach shaves 90% of the time off this check:

```tsx
  for (let i = 0; i < vertices.length; i++) {
    const v = vertices[i];
    const svc = v.service.toLowerCase();
    const op = v.operation.toLowerCase();
    for (let j = 0; j < uiFindArr.length; j++) {
      const tx = uiFindArr[j];
      if (svc.includes(tx) || op.includes(tx)) vertexSet.add(v);
    }
  }
```

Reply:
what makes it that much faster?
I had initially combined service and operation into `possibleMatch` as I wanted to use `matchSorter` to match across both with fancier matching rules than `includes` (but ultimately it'd take a fair deal of work... still think there's some value in that at some point)

Reply:
I split into `svc` & `op` with two includes checks. are `for` loops faster than `forEach` and `some`? without a `break` in the inner `for` loop it seems like `some` should be faster?

Reply:
I'm thinking it's the use of `.forEach`, `.some` and the construction of a new string.

You're right, adding a `break` into the inner loop would have improved the perf even more.

I explored the matter via `console.time` and `console.timeEnd`.

Reply:
that's surprising. I updated to `for` loops